snuba/web/db_query.py

```diff
@@ -607,3 +615,141 @@ def raw_query(
                 "experiments": clickhouse_query.get_experiments(),
             },
         )
+
+
+def _get_allocation_policy(
+    clickhouse_query: Union[Query, CompositeQuery[Table]]
+) -> AllocationPolicy:
+    """given a query, find the allocation policy in its from clause, in the case
+    of CompositeQuery, follow the from clause until something is querying from a table
+    and use that table's allocation policy.
+
+    **GOTCHAS**
+        - Does not handle joins, will return PassthroughPolicy
+        - In case of error, returns PassthroughPolicy, fails quietly (but logs to sentry)
+    """
+    from_clause = clickhouse_query.get_from_clause()
+    if isinstance(from_clause, Table):
+        return from_clause.allocation_policy
+    elif isinstance(from_clause, ProcessableQuery):
+        return _get_allocation_policy(cast(Query, from_clause))
+    elif isinstance(from_clause, CompositeQuery):
+        return _get_allocation_policy(from_clause)
+    elif isinstance(from_clause, JoinClause):
+        # HACK (Volo): Joins are a weird case for allocation policies and we don't
+        # actually use them anywhere so I'm purposefully just kicking this can down the
+        # road
+        return DEFAULT_PASSTHROUGH_POLICY
+    else:
+        logger.exception(
+            f"Could not determine allocation policy for {clickhouse_query}"
```

Comment:
Could certain useful info from the Clickhouse query be passed as tags? I'm guessing this message would get truncated and we might miss the from clause.